### PR TITLE
ENH: Allow inplace copying in place in "detrend" function

### PR DESCRIPTION
**Pull request by OGordon100**
Addresses issue #9790, and allows detrend to potentially halve memory consumption. I have not added this argument to the other functions in the file, as only in detrend is the variable overwritten by a copy of itself. In most cases I believe this to be a needless use of memory.

### CODE REVIEW

File: scipy/signal/signaltools.py [hunk 1]
```diff
@@ -2525,6 +2525,8 @@ def detrend(data, axis=-1, type='linear', bp=0):
         A sequence of break points. If given, an individual linear fit is
         performed for each part of `data` between two break points.
         Break points are specified as indices into `data`.
+    copy : boolean, optional
+        If False, try to avoid a copy and do inplace detrending instead. This may reduce memory usage. Default true
```
Comment (larsoner) [inline scipy/signal/signaltools.py]: Can you PEP8-ify this with < 80 char lines?

**Review by larsoner — COMMENTED** [review]
Other than a PEP8 complaint, looks reasonable to me. But could you add a test that `copy=False` does indeed modify in place? You could just do:
```
x = np.ones(10)
detrend(x, copy=False)
assert_array_equal(x, np.zeros(10))
```

**Comment by ewmoore** [comment]
1.  No reason not to also modify the `type='constant'` path too.
2.  Probably should name the keyword `overwrite_data` to mirror the name used for this feature elsewhere, e.g. in `linalg`.
3. Are you sure that this is saving any memory?  A plain `signal.detrend(x)` call will allocate at least the Vandermonde matrix needed to do the fitting (2x the data input) and the computed trend (1x the data input).  Additional memory will also certainly be allocated in `linalg.lstsq`.  One can contrast this with, for instance, `linalg.lu`, where if the input is consumed, no additional memory need be allocated at all.

**Comment by OGordon100** [comment]
Agreed. Can't believe I didn't notice that!

I've just noticed that detrend is also not unit tested at all. I'll add a few to the request ASAP.

File: scipy/signal/signaltools.py (hunk 1 above)
Comment (OGordon100) [reply scipy/signal/signaltools.py]: Done.

I did what ewmoore suggested and changed the name from "copy" to "overwrite_data." This also meant swapping the logic around, so copy=True equates to overwrite_data=False.

I've also added in a couple of simple unit tests, and one that directly compares overwrite_data=True and overwrite_data=False.

File: scipy/signal/signaltools.py [hunk 2]
```diff
@@ -2505,7 +2505,7 @@ def vectorstrength(events, period):
     return strength, phase
 
 
-def detrend(data, axis=-1, type='linear', bp=0):
+def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=True):
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: you cannot change the current behavior, so default should be False

File: scipy/signal/signaltools.py [hunk 3]
```diff
@@ -2525,6 +2525,9 @@ def detrend(data, axis=-1, type='linear', bp=0):
         A sequence of break points. If given, an individual linear fit is
         performed for each part of `data` between two break points.
         Break points are specified as indices into `data`.
+    overwrite_data : boolean, optional
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: `bool, optional`

File: scipy/signal/signaltools.py [hunk 4]
```diff
@@ -2525,6 +2525,9 @@ def detrend(data, axis=-1, type='linear', bp=0):
...
+        If True, try to avoid a copy and do inplace detrending instead.
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: "try to avoid" --> "avoid"

File: scipy/signal/signaltools.py [hunk 5]
```diff
@@ -2525,6 +2525,9 @@ def detrend(data, axis=-1, type='linear', bp=0):
...
+        This may reduce memory usage. Default false
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: why "may" - it will right?

Comment (rgommers) [reply scipy/signal/signaltools.py]: "Default is False"

File: scipy/signal/tests/test_signaltools.py [hunk 6]
```diff
@@ -184,15 +183,15 @@ def test_convolve_method(self, n=100):
...
+                for mode in ['valid', 'full', 'same']]
 
         # These are random arrays, which means test is much stronger than
         # convolving testing by convolving two np.ones arrays
         np.random.seed(42)
         array_types = {'i': np.random.choice([0, 1], size=n),
                        'f': np.random.randn(n)}
         array_types['b'] = array_types['u'] = array_types['i']
-        array_types['c'] = array_types['f'] + 0.5j*array_types['f']
+        array_types['c'] = array_types['f'] + 0.5j * array_types['f']
```
Comment (rgommers) [inline scipy/signal/tests/test_signaltools.py]: this change is incorrect, the operator with higher precedence can be written without spaces

File: scipy/signal/tests/test_signaltools.py [hunk 7]
```diff
@@ -224,16 +223,16 @@ def test_convolve_method_large_input(self):
         # This is really a test that convolving two large integers goes to the
         # direct method even if they're in the fft method.
         for n in [10, 20, 50, 51, 52, 53, 54, 60, 62]:
-            z = np.array([2**n], dtype=np.int64)
+            z = np.array([2 ** n], dtype=np.int64)
```
Comment (rgommers) [inline scipy/signal/tests/test_signaltools.py]: all changes to `**` are wrong, we don't want spaces around `**`, please undo everywhere

File: scipy/signal/tests/test_signaltools.py [hunk 8]
```diff
@@ -2654,3 +2655,17 @@ def test_basic(self):
...
+class TestDetrend(object):
+
+    def test_basic(self):
+        detremded = detrend(array([1, 2, 3]))
+        detrended_exact = array([0, 0, 0])
+        assert_array_almost_equal(detremded, detrended_exact)
+
+    def test_copy(self):
+        x = np.ones(10)
+        overwritten = detrend(x, overwrite_data=True)
```
Comment (rgommers) [inline scipy/signal/tests/test_signaltools.py]: this test is incorrect, and only happens to pass because `x` has slope zero. you need to do the non-inplace detrend first. can you change `x` so this fails, and then swap the order of the detrends?

**Review by rgommers — CHANGES_REQUESTED** [review]
I pointed out a number of issues. Main comments:

- I suggest to discard the PEP8 changes, since half are incorrect and the rest is not necessary either (it's not code needed for this PR).
- Similar to @ewmoore, I'm not yet sold on this being a good idea. Can you please measure how much memory this saves? If it's 50% then yes this makes sense, but it is clearly less.

File: scipy/signal/tests/test_signaltools.py (hunk 7 above)
Comment (OGordon100) [reply scipy/signal/tests/test_signaltools.py]: Apologies for this. I committed straight from pycharm with optimise code on.

File: scipy/signal/signaltools.py (hunk 2 above)
Comment (OGordon100) [reply scipy/signal/signaltools.py]: Got mixed up when I changed it from "copy" to "overwrite_data." It shouldn't have changed! You're right.

File: scipy/signal/tests/test_signaltools.py (hunk 8 above)
Comment (OGordon100) [reply scipy/signal/tests/test_signaltools.py]: I've added some semi-random numbers to give a slope not equal to zero, and swapped the order around. I also fixed a typo I made in the test done previous, and reverted the spacing changes pycharm made when.

**Comment by OGordon100** [comment]
To comment on why I feel the change is needed, I have copied the function with/without the .copy command into two files, and profiled memory usage with the [memory profiler](https://pypi.org/project/memory-profiler/) package.

`Line #    Mem usage    Increment   Line Contents
================================================
    26   32.648 MiB   32.648 MiB   @profile
    27                             def detrend_test():
    28
    29                                 # Set defaults
    30   32.648 MiB    0.000 MiB       axis=-1
    31   32.648 MiB    0.000 MiB       bp=0
    32   32.648 MiB    0.000 MiB       type='linear'
    33
    34                                 # Make data
    35  223.207 MiB  190.559 MiB       data = np.random.rand(5000, 5000)
    36
    37                                 # Detrend function
    38  223.207 MiB    0.000 MiB       if type not in ['linear', 'l', 'constant', 'c']:
    39                                     raise ValueError("Trend type must be 'linear' or 'constant'.")
    40  223.207 MiB    0.000 MiB       data = asarray(data)
    41  223.207 MiB    0.000 MiB       dtype = data.dtype.char
    42  223.207 MiB    0.000 MiB       if dtype not in 'dfDF':
    43                                     dtype = 'd'
    44  223.207 MiB    0.000 MiB       if type in ['constant', 'c']:
    45                                     ret = data - expand_dims(mean(data, axis), axis)
    46                                     return ret
    47                                 else:
    48  223.207 MiB    0.000 MiB           dshape = data.shape
    49  223.207 MiB    0.000 MiB           N = dshape[axis]
    50  223.207 MiB    0.000 MiB           bp = sort(unique(r_[0, bp, N]))
    51  223.207 MiB    0.000 MiB           if np.any(bp > N):
    52                                         raise ValueError("Breakpoints must be less than length "
    53                                                          "of data along given axis.")
    54  223.207 MiB    0.000 MiB           Nreg = len(bp) - 1
    55                                     # Restructure data so that axis is along first dimension and
    56                                     #  all other dimensions are collapsed into second dimension
    57  223.207 MiB    0.000 MiB           rnk = len(dshape)
    58  223.207 MiB    0.000 MiB           if axis < 0:
    59  223.207 MiB    0.000 MiB               axis = axis + rnk
    60  223.207 MiB    0.000 MiB           newdims = r_[axis, 0:axis, axis + 1:rnk]
    61  223.207 MiB    0.000 MiB           newdata = reshape(transpose(data, tuple(newdims)),
    62  223.207 MiB    0.000 MiB                             (N, _prod(dshape) // N))
    63                                     ##########################################################
    64  414.090 MiB  190.883 MiB           newdata = newdata.copy()  # make sure we have a copy
    65                                     ##########################################################
    66  414.090 MiB    0.000 MiB           if newdata.dtype.char not in 'dfDF':
    67                                         newdata = newdata.astype(dtype)
    68                                     # Find leastsq fit and remove it for each piece
    69  609.797 MiB    0.000 MiB           for m in range(Nreg):
    70  414.090 MiB    0.000 MiB               Npts = bp[m + 1] - bp[m]
    71  414.348 MiB    0.258 MiB               A = ones((Npts, 2), dtype)
    72  414.348 MiB    0.000 MiB               A[:, 0] = cast[dtype](arange(1, Npts + 1) * 1.0 / Npts)
    73  414.348 MiB    0.000 MiB               sl = slice(bp[m], bp[m + 1])
    74  609.293 MiB  194.945 MiB               coef, resids, rank, s = linalg.lstsq(A, newdata[sl])
    75  609.797 MiB    0.504 MiB               newdata[sl] = newdata[sl] - dot(A, coef)
    76                                     # Put data back in original shape.
    77  609.797 MiB    0.000 MiB           tdshape = take(dshape, newdims, 0)
    78  609.797 MiB    0.000 MiB           ret = reshape(newdata, tuple(tdshape))
    79  609.797 MiB    0.000 MiB           vals = list(range(1, rnk))
    80  609.797 MiB    0.000 MiB           olddims = vals[:axis] + [0] + vals[axis:]
    81  609.797 MiB    0.000 MiB           ret = transpose(ret, tuple(olddims))
    82
    83  609.797 MiB    0.000 MiB           return ret

`

Compare this to

`
Line #    Mem usage    Increment   Line Contents
================================================
    26   32.547 MiB   32.547 MiB   @profile
    27                             def detrend_test():
    28
    29                                 # Set defaults
    30   32.547 MiB    0.000 MiB       axis=-1
    31   32.547 MiB    0.000 MiB       bp=0
    32   32.547 MiB    0.000 MiB       type='linear'
    33
    34                                 # Make data
    35  223.172 MiB  190.625 MiB       data = np.random.rand(5000, 5000)
    36
    37                                 # Detrend function
    38  223.172 MiB    0.000 MiB       if type not in ['linear', 'l', 'constant', 'c']:
    39                                     raise ValueError("Trend type must be 'linear' or 'constant'.")
    40  223.172 MiB    0.000 MiB       data = asarray(data)
    41  223.172 MiB    0.000 MiB       dtype = data.dtype.char
    42  223.172 MiB    0.000 MiB       if dtype not in 'dfDF':
    43                                     dtype = 'd'
    44  223.172 MiB    0.000 MiB       if type in ['constant', 'c']:
    45                                     ret = data - expand_dims(mean(data, axis), axis)
    46                                     return ret
    47                                 else:
    48  223.172 MiB    0.000 MiB           dshape = data.shape
    49  223.172 MiB    0.000 MiB           N = dshape[axis]
    50  223.172 MiB    0.000 MiB           bp = sort(unique(r_[0, bp, N]))
    51  223.172 MiB    0.000 MiB           if np.any(bp > N):
    52                                         raise ValueError("Breakpoints must be less than length "
    53                                                          "of data along given axis.")
    54  223.172 MiB    0.000 MiB           Nreg = len(bp) - 1
    55                                     # Restructure data so that axis is along first dimension and
    56                                     #  all other dimensions are collapsed into second dimension
    57  223.172 MiB    0.000 MiB           rnk = len(dshape)
    58  223.172 MiB    0.000 MiB           if axis < 0:
    59  223.172 MiB    0.000 MiB               axis = axis + rnk
    60  223.172 MiB    0.000 MiB           newdims = r_[axis, 0:axis, axis + 1:rnk]
    61  223.172 MiB    0.000 MiB           newdata = reshape(transpose(data, tuple(newdims)),
    62  223.172 MiB    0.000 MiB                             (N, _prod(dshape) // N))
    63                                     ##########################################################
    64  223.172 MiB    0.000 MiB           newdata = newdata  # don't copy
    65                                     ##########################################################
    66  223.172 MiB    0.000 MiB           if newdata.dtype.char not in 'dfDF':
    67                                         newdata = newdata.astype(dtype)
    68                                     # Find leastsq fit and remove it for each piece
    69  419.047 MiB    0.000 MiB           for m in range(Nreg):
    70  223.172 MiB    0.000 MiB               Npts = bp[m + 1] - bp[m]
    71  223.586 MiB    0.414 MiB               A = ones((Npts, 2), dtype)
    72  223.586 MiB    0.000 MiB               A[:, 0] = cast[dtype](arange(1, Npts + 1) * 1.0 / Npts)
    73  223.586 MiB    0.000 MiB               sl = slice(bp[m], bp[m + 1])
    74  418.535 MiB  194.949 MiB               coef, resids, rank, s = linalg.lstsq(A, newdata[sl])
    75  419.047 MiB    0.512 MiB               newdata[sl] = newdata[sl] - dot(A, coef)
    76                                     # Put data back in original shape.
    77  419.047 MiB    0.000 MiB           tdshape = take(dshape, newdims, 0)
    78  419.047 MiB    0.000 MiB           ret = reshape(newdata, tuple(tdshape))
    79  419.047 MiB    0.000 MiB           vals = list(range(1, rnk))
    80  419.047 MiB    0.000 MiB           olddims = vals[:axis] + [0] + vals[axis:]
    81  419.047 MiB    0.000 MiB           ret = transpose(ret, tuple(olddims))
    82
    83  419.047 MiB    0.000 MiB           return ret
`

I was mistaken in saying it was half, but for this semi-large 5000x5000 array it's about 1/3. For small arrays it's a smaller percentage, yes, but an improvement nonetheless. For my personal work with large arrays a few gigabytes in size, this is the difference between running and crashing with MemoryError.

I also refer you to the [core python documentation for .copy()](https://docs.python.org/3.7/library/copy.html) - it explains quite well why the .copy() command causes increased memory usage:

> Assignment statements in Python do not copy objects, they create bindings between a target and an object. For collections that are mutable or contain mutable items, a copy is sometimes needed so one can change one copy without changing the other. 

Or in other words, .copy just what you think it does - makes a copy of the object in memory, but leaves the old one behind. For the detrend function we are making an exact duplicate of the input data in RAM, but never using the old copy.

I also refer you to [np.array](https://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html) and [sklearn preprocessing](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html). The source of this function implements the "copy" argument in the exact same way as I have proposed here. I also refer you to numerous, highly similar discussions for scikit learn [[1]](urlhttps://github.com/scikit-learn/scikit-learn/pull/11973 [[2]](https://github.com/scikit-learn/scikit-learn/issues/11970) [[3]](https://github.com/scikit-learn/scikit-learn/pull/11966#issuecomment-417913111) [[4]](https://github.com/scikit-learn/scikit-learn/pull/11973).

Hopefully this explains my reasoning!

**Review by OGordon100 — COMMENTED** [review]
Made changes as asked. Discussed in comment.

**Comment by rgommers** [comment]
Thanks for the profiling results @OGordon100. 1/3 memory use seems enough to me to justify adding a keyword for not copying.

> I also refer you to numerous, highly similar discussions for scikit learn [1] [2] [3] [4].

Note that those are different, they don't require adding a new keyword, it's just memory saving under the hood (which is always good, no downsides).

File: scipy/signal/signaltools.py [hunk 9]
```diff
@@ -2567,7 +2569,8 @@ def detrend(data, axis=-1, type='linear', bp=0):
         newdims = r_[axis, 0:axis, axis + 1:rnk]
         newdata = reshape(transpose(data, tuple(newdims)),
                           (N, _prod(dshape) // N))
-        newdata = newdata.copy()  # make sure we have a copy
+        if overwrite_data is False:
```
Comment (rgommers) [inline scipy/signal/signaltools.py]: should be `if not overwrite_data`:

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
LGTM now, module the last minor comment. Would be good if you could squash your commits when addressing that.

**Comment by rgommers** [comment]
Merged, thanks @OGordon100
